docs/DesignDocs/ExtendedEnclaveInitializationData_V2.md

```diff
@@ -0,0 +1,182 @@
+
+For EEID enclaves with heap/stack size and #TCS selected by the loader at enclave loading time, the loader adds heap pages, thread sections and the EEID page after the base enclave's code and initialized data are measured by the CPU. The `hash_state` records the measurement before the heap and thread sections are added. This type of of EEID enclave's Enclave_Size in `SECS` is fixed as a large value, for example 64GB.
+
+As discussed in the original EEID design doc(URL), an EEID enclave with heap/stack size and #TCS selected by the loader at enclave load time needs to have one thread section in the base image to capture the expected `TCS.OENTRY` value in the base enclave measurement, which does complicate the base enclave measurement recreation algorithm used in EEID enclave attestation verification. This updated design addresses the need to capture the expected `TCS.OENTRY` value in a different way.
+
+`oe_sgx_enclave_properties` is part of the initialized data of the SGX enclave, and covered by the base enclave measurement. The enclave loader will include the expected `TCS.OENTRY` value in `oe_sgx_enclave_properties`. The definition of `oe_sgx_enclave_properties` is changed from:
+
+```C
+/* Extends oe_enclave_properties_header_t base type */
+typedef struct _oe_sgx_enclave_properties
+{
+    /* (0) */
+    oe_enclave_properties_header_t header;
+
+    /* (32) */
+    oe_sgx_enclave_config_t config;
+
+    /* (48) */
+    oe_sgx_enclave_image_info_t image_info;
+
+    /* (96)  */
+    uint8_t sigstruct[OE_SGX_SIGSTRUCT_SIZE];
+
+    /* (1904) end-marker to make sure 0-filled sigstruct doesn't get omitted */
+    uint64_t end_marker;
+} oe_sgx_enclave_properties_t;
+```
+
+to:
+
+```C
+/* Extends oe_enclave_properties_header_t base type */
+typedef struct _oe_sgx_enclave_properties
+{
+    /* (0) */
+    oe_enclave_properties_header_t header;
+
+    /* (32) */
+    oe_sgx_enclave_config_t config;
+
+    /* (48) */
+    oe_sgx_enclave_image_info_t image_info;
+
+    /* (96)  */
+    uint8_t sigstruct[OE_SGX_SIGSTRUCT_SIZE];
+
+    /* (1904) from oeimage.entry_rva, must match TCS.OENTRY */
+    uint64_t entry_rva;
+} oe_sgx_enclave_properties_t;
+```
+
+The EEID initialization code inside the enclave makes sure the `entry_point` in the EEID data structure matches `entry_rva` in `oe_sgx_enclave_properties`.
+
+With this change, the base enclave of an EEID enclave with heap/stack size and #TCS selected by the loader can safely use  `NumStackPages=0`, `NumHeapPages=0`, and `NumTCS=0`. Recording the expected `TCS.OENTRY` value in `oe_sgx_enclave_properties` of a SGX enclave (regular enclave or EEID enclave) also simplifies the implementation of Enclave Dynamic Memory Management (EDMM) based thread context creation support in the future, where the enclave code needs to verify the content of a TCS page added by the OS.
```

Comment:
If the base enclave has no TCS, without including the entry point in the initialized data (measured) and checking `entry_point` in EEID data against it, a malicious loader can set `entry_point` to a different value than the enclave developer intended, and add TCS page(s) with the spoofed value. The EEID attestation verifier will see all measurement checks pass.

Reply:
I must be missing something here. With static memory sizes this is not a problem (at signing or creation time, as long as numTCS > 0). With dynamic memory settings, the final measurement (resigned, at creation time) does include the TCS's (and numTCS > 0). So in both settings, if the host puts a different entry point into the TCSs, attestation would fail, because the verifier expects the EEID entry point (see `oe_remeasure_memory_pages`). The base image hash would match (with numTCS=0), but the extended hash would not. numTCS is also in the EEID settings, maybe the verifier should check that it's not 0?